### PR TITLE
student view team: Show team member profile photo #5943

### LINKED ISSUE #5943: student view team: Show team member profile photo

### PR DESCRIPTION
**Pull request by tarun14110**
Fixes #5943

screenshot from 2016-08-01 00 26 23

### CODE REVIEW

**Comment by tarun14110** [comment]
ready for review.

**Comment by chowyb** [comment]
Confirmed that the feature works for non-default profile pictures:

pp

Would it be possible to add tests such as checking for whether the correct emails/pictures are displayed? It's better not to rely solely on GodMode to test this feature.

**Comment by damithc** [comment]
One option: I think we have a test somewhere that tests for uploading profile pictures. We can add a test in the same class to load the team page and ensure the uploaded photo is visible?.

**Comment by tarun14110** [comment]
@damithc I think I didn't get that clearly. We want to ensure the uploaded photo to be visible to all his team members on `viewTeam` page in the course?

**Comment by damithc** [comment]
Yes. Let student 1 upload the photo and ensure the photo is visible in the 'view team' page of a team member of that student.

**Comment by tarun14110** [comment]
I think we don't have `upload photo` tests other than testing through comparing `HTML` files in `StudentProfilePageUiTest`. `StudentProfileEditSaveActionTest` is just testing saving of other fields other than pictures. `StudentProfilePictureActionTest` tests retrieval of picture from given params.

@chowyb we are checking displayed teammates in `StudentCourseDetailsPageActionTest`. Isn't that enough to ensure correct teammates are displayed. And to check for displayed content, we are using `HTML` comparisons ?

Otherwise I can add a test in `StudentCourseDetailsPageActionTest`, as discussed above?

**Comment by chowyb** [comment]
@tarun14110 yes, while testing through comparing HTML might work now, it is not too robust as changes to the page later on may easily overwrite the test HTML pages via GodMode, which may slip by the person reviewing that change. I would prefer adding code to the UI test such as (as an example; note that the following methods may need to be implemented as well):

```
assertEquals("charlie.d.tmms@gmail.tmt", getEmailOfStudent(1));
assertEquals("Charlie Davis", getNameOfStudent(1));
assertEquals("page/studentProfilePic...", getPictureOfStudent(1));
```

This ensures that the picture is associated with the correct student. To handle this, there is the option given above: to add a test within `StudentProfilePageUiTest` after uploading a profile picture to check that the student's teammates are able to see the new profile picture. Another possible method to test for this would be to add a profile picture to the input test data in `StudentCourseDetailsPageUiTest`, and test that the image is shown, but I suspect this might be a more difficult option.

**Comment by tarun14110** [comment]
@chowyb added test, again ready for your review.

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 1]
```diff
@@ -575,6 +575,15 @@ Apply to thead to make headers not bold.
     min-width: 130px;
 }
 
+.teamMembersDetailsCell{
+    vertical-align: middle; 
+    padding-left:30px;
```
Comment (chowyb) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Add a space after the colon here.

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 2]
```diff
@@ -575,6 +575,15 @@ Apply to thead to make headers not bold.
     min-width: 130px;
 }
 
+.teamMembersDetailsCell{
```
Comment (chowyb) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Add a space just before the `{` here.

File: src/main/webapp/stylesheets/teammatesCommon.css [hunk 3]
```diff
@@ -575,6 +575,15 @@ Apply to thead to make headers not bold.
...
+    vertical-align: middle; 
+    padding-left:30px;
+}
+
+.teamMembersPhotoCell{
+    padding-bottom:5px;
```
Comment (chowyb) [inline src/main/webapp/stylesheets/teammatesCommon.css]: Same issues here (space before `{`, and space after `:`)

**Comment by chowyb** [comment]
Some styling changes requested.

File: src/main/webapp/js/common.js [hunk 4]
```diff
@@ -179,6 +181,19 @@ $(document).on('ajaxComplete ready', function() {
 });
 
 /**
+ * @param elements:
+ * identifier that points to elements with
+ * class: profile-pic-icon-click or profile-pic-icon-hover
+ */
```
Comment (damithc) [inline src/main/webapp/js/common.js]: Follow the normal javadoc comment format.
Do not mention where the function is used. A function should be independent of the call site.

**Comment by damithc** [comment]
Minor cosmetic change requested.

File: src/main/webapp/js/common.js [hunk 5]
```diff
@@ -179,6 +181,19 @@ $(document).on('ajaxComplete ready', function() {
 });
 
 /**
+ * @param elements:
+ *     identifier that points to elements for which child 'img' tags are
+ *     checked for error and assigning default value in case of error.
```
Comment (damithc) [inline src/main/webapp/js/common.js]: Something like this?
```java
/**
  * Binds a default image if the image is missing.
  * @param elements A collection of image elements. 
  */
```
The function name can be changed too. e.g. `bindDefaultImageIfMissing`
Furthermore, this function makes more sense if written for a single element and then the call site can call it in a loop.
The default image can be a parameter.

**Comment by tarun14110** [comment]
@chowyb updated,  please have look.

File: src/main/webapp/js/instructor.js [hunk 6]
```diff
@@ -136,19 +135,6 @@ function setupFsCopyModal() {
 /**
  * @param elements:
  * identifier that points to elements with
- * class: profile-pic-icon-click or profile-pic-icon-hover
- */
-function bindErrorImages(elements) {
```
Comment (wkurniawan07) [inline src/main/webapp/js/instructor.js]: This function is called in a few other places, and renaming it in `common.js` is what causes the Travis build to stall.

**Comment by damithc** [comment]
@tarun14110 Let's merge this soon. I need it in my courses.

**Comment by wkurniawan07** [comment]
There are some parts of this PR that I want to comment on, but if the feature is urgently needed, they can be addressed on another issue/PR. @damithc @chowyb thoughts?

**Comment by damithc** [comment]
@wkurniawan07  go ahead. It's fine if we can release the feature within a week.

File: src/test/java/teammates/test/cases/ui/StudentProfileEditSaveActionTest.java [hunk 7]
```diff
@@ -8,12 +8,17 @@
...
+import teammates.common.datatransfer.StudentAttributes;
 import teammates.common.datatransfer.StudentProfileAttributes;
+import teammates.common.exception.EntityDoesNotExistException;
 import teammates.common.util.Const;
 import teammates.common.util.FieldValidator;
 import teammates.common.util.StringHelper;
 import teammates.test.driver.AssertHelper;
 import teammates.ui.controller.RedirectResult;
+import teammates.ui.controller.ShowPageResult;
+import teammates.ui.controller.StudentCourseDetailsPageAction;
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/ui/StudentProfileEditSaveActionTest.java]: I don't really like having the dependency to `StudentCourseDetailsPageAction` here. This test should only be testing `StudentProfileEditSaveAction` as the test name suggests. What is being tested here and why is this necessary?

Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/ui/StudentProfileEditSaveActionTest.java]: @tarun14110

Comment (tarun14110) [inline src/test/java/teammates/test/cases/ui/StudentProfileEditSaveActionTest.java]: We are testing that if a user uploaded his profile pic, that should be visible on his/her team member's `teamView` page. The reason is as stated [above](https://github.com/TEAMMATES/teammates/pull/5948#issuecomment-237472747) by chowyb.

Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/ui/StudentProfileEditSaveActionTest.java]: @chowyb's suggestion is using UI tests, not action tests. It is not right to have `StudentCourseDetailsPageAction` appearing in an action test not related to it. If you still want to test via action test, go to `StudentCourseDetailsPageActionTest`, upload a profile picture before the test is executed (e.g at `@BeforeClass`), and test accordingly.

Comment (tarun14110) [inline src/test/java/teammates/test/cases/ui/StudentProfileEditSaveActionTest.java]: @wkurniawan07 sure, will do that.

**Comment by damithc** [comment]
@tarun14110 Let's finish this. I need this to be deployed by next Tuesday.

**Comment by tarun14110** [comment]
Sorry for being late. @wkurniawan07 @chowyb I updated the tests, Can you please review.

**Comment by wkurniawan07** [comment]
This is not yet satisfactory, but in the interest of time I guess this can be merged first? The feature is working as far as I (and @chowyb) can see. I'll do the necessary test fixes myself @damithc 
For future reference I'll still comment on what needs to be worked on.

File: src/test/java/teammates/test/cases/ui/StudentCourseDetailsPageActionTest.java [hunk 8]
```diff
@@ -112,8 +117,66 @@ public void testExecuteAndPostProcess() {
...
+        
+        StudentCourseDetailsPageAction pageAction = getAction(submissionParam);
+        ShowPageResult pageResult = getShowPageResult(pageAction);
+        StudentCourseDetailsPageData pageData = (StudentCourseDetailsPageData) pageResult.data;
+        
+        List<StudentAttributes> actualStudentsList = pageData.getStudentCourseDetailsPanel().getTeammates();
+        boolean isStudentDisplayedOnViewTeam = false;
+        for (StudentAttributes stud : actualStudentsList) {
+            if (student1.email.equals(stud.email) && student1.name.equals(stud.name)
+                    && student1.getPublicProfilePictureUrl().equals(stud.getPublicProfilePictureUrl())) {
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/ui/StudentCourseDetailsPageActionTest.java]: If it is done this way, we cannot be sure that there actually exists a non-default profile picture URL. The URL should be hardcoded here.

File: src/test/java/teammates/test/cases/ui/StudentCourseDetailsPageActionTest.java [hunk 9]
```diff
@@ -17,6 +21,7 @@
 import teammates.ui.controller.ShowPageResult;
 import teammates.ui.controller.StudentCourseDetailsPageAction;
 import teammates.ui.controller.StudentCourseDetailsPageData;
+import teammates.ui.controller.StudentProfileEditSaveAction;
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/ui/StudentCourseDetailsPageActionTest.java]: Same mistake occurs: an irrelevant Action class being imported in the action test not related to it.

File: src/test/java/teammates/test/cases/ui/StudentCourseDetailsPageActionTest.java [hunk 10]
```diff
@@ -112,8 +117,66 @@ public void testExecuteAndPostProcess() {
...
+    @Test
+    public void testTeamMemberDetailsOnViewTeamPage() throws EntityDoesNotExistException {
+        AccountAttributes student = dataBundle.accounts.get("student1InCourse1");
+        
+        String[] submissionParams = createValidParamsForProfile();
+        StudentProfileAttributes expectedProfile = getProfileAttributesFrom(submissionParams);
+        gaeSimulation.loginAsStudent(student.googleId);
+        
+        // adding profile picture for student1InCourse1
+        StudentProfileEditSaveAction action = getStudentProfileEditSaveAction(submissionParams);
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/ui/StudentCourseDetailsPageActionTest.java]: This is not adding profile picture; this is adding student profile which, while may be good, is not our intention here.
